### PR TITLE
Tag editor modal

### PR DESCRIPTION
**Pull request by jjlangholtz**
Add `TagEditorModal` component to be used for tagging services, catalogs, orchestration templates, etc. The tag editor allows assignment and un-assignment of tags for resources individually or in bulk. Assignment/un-assignment for a large number of resources is inefficient until a bulk tagging API is implemented.

UI improvements mentioned in manageiq/manageiq-design#6 will be implemented in a follow-up PR.

Bulk editing:
screen shot 2016-12-19 at 5 16 51 pm

Individual resource editing:
screen shot 2016-12-19 at 5 16 27 pm

@miq-bot add_label euwe/no, enhancement

https://www.pivotaltracker.com/story/show/135042907

### CODE REVIEW

**Comment by chriskacerguis** [comment]
@cshinn @serenamarie125 @beanh66 for your UX review :)

**Comment by serenamarie125** [comment]
@jjlangholtz how is Edit Tags invoked here, is it in a button group and/or available in the kebab ?
and what does reset do?  Thanks!

**Comment by beanh66** [comment]
@jjlangholtz Looks good! I think the + button needs to move over to the left to be right next to the second drop down box. Are you thinking we will wait for the UI improvements follow up PR to get rid of the + button? That makes sense, just wanted to confirm :)

**Comment by jjlangholtz** [comment]
@serenamarie125 Edit Tags is invoked via the kebab menu when viewing individual services and through the 'Policy' middle button of the explorer view button group. The reset button changes the state of the tags to be the same as when the modal was first opened.

@beanh66 I'm planning to make all of the UI improvements, including the removal of the + button in a separate PR.

File: client/app/states/services/details/details.html [hunk 1]
```diff
@@ -79,6 +79,7 @@
                   <!--<a href="#" title="Suspend this service" ng-click="vm.handlePowerOperation('suspend', vm.service)">{{'Suspend'|translate}}</a>-->
                 <!--</li>-->
                 <li><a href="#" title="Edit this service" ng-click="vm.editServiceModal()" ng-show="vm.showEditService">{{'Edit'|translate}}</a></li>
+                <li><a href="#" title="Edit tags" ng-click="vm.tagEditorModal()" ng-show="vm.showTagEditor">{{'Edit Tags'|translate}}</a></li>
```
Comment (chalettu) [inline client/app/states/services/details/details.html]: Should we try and translate the title tags as well?

Comment (jjlangholtz) [inline client/app/states/services/details/details.html]: Good call out, I think any user visible string should probably be translated. Probably worthy of a larger discussion, but I'm thinking we should be more consistent with our use of the `title` attribute in general. For consistency purposes we should probably not use `title` directly and instead use `uib-tooltip`s.

**Comment by dtaylor113** [comment]
Hi, cloned this branch and ran through blueprint tagging and 'items on canvas' tagging -all looks great!  Nice work @jjlangholtz :+1:

**Comment by chriskacerguis** [comment]
@serenamarie125 @beanh66 @cshinn could you please add your approval or feedback before merge?

File: client/app/services/tagging.service.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    // queries all available tags for that resource. The result is filtered of
+    // invalid tags (missing required properties).
+    function queryAvailableTags(resourceUrl) {
+      var queryOptions = {
+        expand: 'resources',
+        attributes: ['categorization', 'category'],
+      };
+
+      return CollectionsApi.query(resourceUrl || 'tags', queryOptions)
+        .then(filterValidTags);
```
Comment (AllenBW) [inline client/app/services/tagging.service.js]: but what if this fails!?

**Comment by beanh66** [comment]
@chriskacerguis LGTM!

File: client/app/components/tag-editor-modal/tag-editor-modal-service.factory.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        return tags;
+      }
+    }
+  }
+
+  /** @ngInject */
+  function TagEditorModalController(services, tags, $controller, $uibModalInstance,
+                                    $state, taggingService, EventNotifications) {
+    var vm = this;
+    var base = $controller('BaseModalController', {
```
Comment (AllenBW) [inline client/app/components/tag-editor-modal/tag-editor-modal-service.factory.js]: oh i understand this now, <3 🙇‍♀️

**Comment by chriskacerguis** [comment]
@jjlangholtz @AllenBW what's the word here, is this ready to go?

**Comment by jjlangholtz** [comment]
@chriskacerguis I believe this is ready to go now

**Pull request merged**